javaslang/src/main/java/javaslang/Lazy.java

```diff
@@ -135,14 +133,11 @@ private Lazy(Supplier<? extends T> supplier) {
      */
     @Override
     public T get() {
-        // using a local var speeds up the double-check idiom by 25%, see Effective Java, Item 71
-        Supplier<? extends T> tmp = supplier;
-        if (tmp != null) {
+        if ((value == null) && (supplier != null)) {
```

Comment:
@shipilev, I couldn't find the latest [`jcstress`](http://openjdk.java.net/projects/code-tools/jcstress) on a Maven repo (did find an older one for a [Gradle plugin](https://github.com/jerzykrlk/jcstress-gradle-plugin) though).
I built the project, but it needs [`Java 9`](http://hg.openjdk.java.net/code-tools/jcstress/file/default/pom.xml#l152), and even then I get a compiler warning because  [`JCStressTestProcessor`](http://hg.openjdk.java.net/code-tools/jcstress/file/default/jcstress-core/src/main/java/org/openjdk/jcstress/infra/processors/JCStressTestProcessor.java#l52) has a `@SupportedSourceVersion(SourceVersion.RELEASE_6)` (Javaslang needs `Java 8` and is built with `-werror` via `Maven`).

Do you recommend trying [thread-weaver](https://github.com/google/thread-weaver) instead - though only 18 commits worry me a bit?

ps. The code you commented on was experimental, i.e. trying to make the tests fail - without success, though. The original was (and is again):
```java
transient volatile Supplier<? extends T> supplier;
T value;

T get() {
    if ((value == null) && (supplier != null)) {
        synchronized (this) {
            final Supplier<? extends T> s = supplier;
            if (s != null) {
                value = s.get();
                supplier = null;
            }
        }
    }
    return value;
}
```
i.e. checking the non-volatile first seems to speed up the [benchmarks](https://github.com/javaslang/javaslang/pull/1577/commits/f956463ca578429ef03a999ea194ecc0499f0707#diff-179d7365a1b162ee88ba5aac7bd0c653R46).